### ISSUE TITLE
Using embark recursively during embark--act doesn't work properly

### ISSUE THREAD

**Issue by jakanakaevangeli** [oantolin/embark#121]
To reproduce with emacs -q, run:
```elisp
(load-file "~/src/emacs/embark/embark.el")
(define-key minibuffer-local-map (kbd "M-o") 'embark-act)
```
press `C-x C-f` and then `M-o c` to copy a file. You are now in a minibuffer
window inside `embark--act`. Using embark now (by pressing `M-o o` for example)
will not quit the first minibuffer properly.

Quitting from the inner minnibuffer with `C-g` will also fail to quit the
outer one, but I personally find this very convenient.

Here is my ugly attempt at fixing this. Using a simple unwind-protect turns out
to not be sufficient because the outermost recursive edit would win at
restoring the incorrect window configuration. It would also break the
previously mentioned `C-g` functionality which i like.

```diff
diff --git a/embark.el b/embark.el
index c91e485..35ee212 100644
--- a/embark.el
+++ b/embark.el
@@ -755,6 +755,8 @@ (defun embark--target ()
         (funcall transformer target)
       (cons type target))))
 
+(defvar embark-act--should-unwind nil)
+
 ;;;###autoload
 (defun embark-act (&optional arg)
   "Prompt the user for an action and perform it.
@@ -779,9 +781,18 @@ (defun embark-act (&optional arg)
                                                target)))
     (if (null action)
         (minibuffer-message "Canceled")
-      (embark--act action target)
-      (when (if embark-quit-after-action (not arg) arg)
-        (embark-quit)))))
+      (let ((should-quit
+             (and (minibufferp) (if embark-quit-after-action (not arg) arg))))
+        (setq embark-act--should-unwind nil)
+        (unwind-protect (embark--act action target)
+          (when (and embark-act--should-unwind should-quit)
+            ;; We don't use `embark-quit' here because only the innermost quit
+            ;; sees the correct window configuration to be restored
+            (abort-recursive-edit)))
+        (when should-quit
+          (setq embark-act--should-unwind t)
+          (run-at-time 0 nil #'set 'embark-act--should-unwind nil)
+          (embark-quit))))))
 
 (defun embark--become-keymap ()
   "Return keymap of commands to become for current command."
```

**Comment by oantolin** [comment]
I think this issue is included in the more general issue you reported in #123, right?

**Comment by jakanakaevangeli** [comment]
As I see it, the issues seem to be independent.

This issue is concerned with embark-act and the way it quits from the
minibuffer. Which function it uses to quit is less relevant than the way
it is used. You could replace (embark-quit) with the simpler
(abort-recursive-edit) inside embark-act and this issue would be pretty
much the same, save for the weird window switching that happens inside
embark-quit.

The other issue is concerned with embark-quit itself, how it shouldn't
run the exit hook and its incorrect restoration of window configuration.

**Comment by oantolin** [comment]
Then I don't understand this issue: is it that you want the inner call to embark-act to exit all recursive minibuffers instead of just one level?

**Comment by oantolin** [comment]
The old embark-act did indeed quit all minibuffer and returned you to top-level (it used `(top-level)`, actually). As a recursive minibuffer user, I don't like being returned all the way to top-level and decided on purpose to change `embark-act` to quit only one level.

**Comment by jakanakaevangeli** [comment]
> is it that you want the inner call to embark-act to exit all recursive
> minibuffers instead of just one level?

I want the inner call to embark-act to quit one level and the outer call
to embark-act to quit one level, so two levels in total. Currently the
outer embark-act fails to quit the minibuffer, the call to (embark-quit)
isn't reached because the inner call to embark-act causes a quit
condition.

That is why my patch tries to ensure that (embark-quit) is reached with
the help of unwind-protect.

**Comment by oantolin** [comment]
Oh, I finally understand. Thanks for your patience! Now that I now what the problem is, I wonder if it isn't enough to just run `embark-quit` in an `unwind-protect`, without this should unwind variable.

**Comment by oantolin** [comment]
Um, no probably not.

**Comment by jakanakaevangeli** [comment]
I have found a cleaner solution that doesn't rely on `run-at-time`. The code is
a bit more complicated but in my opinion it is well worth it, getting rid of
`run-at-time` and relying on binding rather than setting the variable.

Using throw instead of abort-recursive-edit is a cool bonus that makes sure the
recursive minibuffer is properly quit even if we are in a `M-x recursive-edit`.

```diff
diff --git a/embark.el b/embark.el
index c91e485..dcc0cd4 100644
--- a/embark.el
+++ b/embark.el
@@ -755,6 +755,9 @@ (defun embark--target ()
         (funcall transformer target)
       (cons type target))))

+(defvar embark-act--in-progress nil
+  "Bound to t if `embark--act' is in progress")
+
 ;;;###autoload
 (defun embark-act (&optional arg)
   "Prompt the user for an action and perform it.
@@ -779,9 +782,25 @@ (defun embark-act (&optional arg)
                                                target)))
     (if (null action)
         (minibuffer-message "Canceled")
-      (embark--act action target)
-      (when (if embark-quit-after-action (not arg) arg)
-        (embark-quit)))))
+      (if (and (minibufferp) (if embark-quit-after-action (not arg) arg))
+          (if (catch 'embark-act
+                (let ((embark-act--in-progress t))
+                  (embark--act action target))
+                nil)
+              ;; An inner embark-act quit with `embark-quit'. We should
+              ;; therefore not quit with `embark-quit' because only the
+              ;; innermost embark-act sees the correct window configuration
+              (if embark-act--in-progress
+                  (throw 'embark-act t)
+                (abort-recursive-edit))
+            ;; We are the inner embark-act, the current window configuration is
+            ;; the correct one, so we should quit with `embark-quit'
+            (if embark-act--in-progress
+                (cl-letf (((symbol-function #'abort-recursive-edit)
+                           (lambda () (throw 'embark-act t))))
+                  (embark-quit))
+              (embark-quit)))
+        (embark--act action target)))))

 (defun embark--become-keymap ()
   "Return keymap of commands to become for current command."
```

**Comment by oantolin** [comment]
What I don't understand is this stuff about only the innermost embark-act seeing the correct window configuration. I thought that the innermost quits, restoring the correct window configuration and leaving that window configuration ready for the second innermost one, so that when it quits it has the correct window configuration also and can restore it leaving it ready for the thirs innermost, and so on.

**Comment by oantolin** [comment]
Oh, I see all the restoring of window configurations runs after the embark-quit calls return. So if minibuffer 1 opens minibuffer 2, you first restore the window configuration of minibuffer 2, then restore the window configuration that was current when you quit minibuffer 1. When you quit minibuffer 1 is after minibuffer 2 was quit, but before it has had a chance to restore the window configuration. So restoring the window configuration for minibuffer 1 gives you an earlier window configuration than the one you wanted. This must be what you meant by "only the innermost quit sees the correct window configuration to be restored".

**Comment by oantolin** [comment]
I'm trying to see if there is a simpler strategy than yours. In particular, I'd love to avoid defining any extra global variables. Does the following patch behave correctly?

```diff
-      (embark--act action target)
-      (when (if embark-quit-after-action (not arg) arg)
-        (embark-quit)))))
+      (unwind-protect
+          (embark--act action target)
+        (when (if embark-quit-after-action (not arg) arg)
```

**Comment by oantolin** [comment]
Sorry, I edited the diff, and I know you've been reading these comments by email, so I don't know if you'll see the correct version.

**Comment by oantolin** [comment]
I think that simple patch I proposed doesn't quite work either. I'm starting to think your solution can't really be simplified that much.

**Comment by oantolin** [comment]
So @jakanakaevangeli, I know you've been hard working very hard on this, but I'm starting to think it's better to just give up on `embark-quit`. In PR #126 @minad asked: is the new quitting behavior worth the complication? To be clear, I'm only wondering about removing `embark-quit`, which is only a part of the new quitting behavior. I changed several things at the same time:

1. Quitting now only quits one level instead of going to the top level. This I like and I'd keep, and is as easy as changing `top-level` to `abort-recursive-edit` in the old code.
2. There is now only `embark-act` with the `C-u` toggle and an `embark-quit-after-action` option. This change seems OK too, I do like saving one keybinding and the conceptual simplicity of having a single command to run actions. I also like that this single command is comfortable for both people that almost always quit after an action and for people that almost never quit after an action. And the custom variable makes it easy for people who really want separate quitting and non-quitting commands to define them.
3. It kind of bothered me that for actions that spawn new windows, if you run then as a non-quitting action, all those windows disappear when you quit or exit the minibuffer at the end, so I decided to try to make `embark-quit`, a command that quits the minibuffer but keeps the window configuration. This is what has proved much trickier to implement correctly than I thought. What I was doing before `embark-quit` was limiting my self to one window-spawning action per minibuffer session and running that action at the very end as a quitting action (or just accepting that the newly spawned windows would disappear). The old way of running quitting actions just quit first and then ran the action, so its window changes didn't get undone.

   I sort of thought that if I had `embark-quit` I'd be happy binding it to `C-g` in the minibuffer and having my window configuration always restored after quitting the minibuffer but I'm not that happy actually: (1) when you exit the minibuffer (press `RET`) the newly spawned windows still disappear, of course, so `embark-quit` isn't a full solution: I don't keep my windows in all cases I want to, but also (2) now if I `C-g` while using `consult-buffer` or other consult commands whose preview involves window changes, I sometimes wind up keeping the window configuration of the preview, which I don't want to keep! So right now, because of  (2), I sometimes quit with `embark-quit` and sometimes with `abort-recursive-edit`; I don't like that this requires planning ahead.
   
All in all, I'm thinking the new `embark-quit` is not worth it: it simplified the implementation of embark-act slightly at the cost of being quite complicated itself (and it is still not fully correct, as your issues and pull requests clearly show @jakanakaevangeli), and it doesn't really fully serve the purpose I had in mind of keeping the window configuration changes made during a minibuffer session in every case I want them and in no case I don't want them.

How would people feel about removing `embark-quit`? (I'd still keep the new `embark-act` with it's `C-u` behavior and the `embark-quit-after-action` variable, and quitting would still exit only one level.) This restoring of the window configuration from before the minibuffer was open is very ingrained in how `read-from-minibuffer` works, and trying to get around it is a struggle I'm not sure is worth it. I feel it's "fighting the Emacs way" and I'd do better to just adapt my workflow to the Emacs way, to accept that window changes while the minibuffer is open are temporary (which, to be fair, is very convenient sometimes).

**Comment by oantolin** [comment]
I'm removing `embark-quit` for now, and going back to the old implementation of quitting actions. 976246cc83b6c619fbc78592252269c7a1103d49
 
This also fixes this issue (please test to make sure it does fix the recursive act case). We can always reintroduce `embark-quit` if we figure out how to implement it nicely. But actually, I'm thinking that this functionality of keeping windows after exiting the minibuffer doesn't really fit in with Embark and might warrant a separate package (which now that you know much more about how to do it than I do, @jakanakaevangeli, you might want to write 😉).

**Comment by minad** [comment]
I should say first that I appreciate thehard work you and @jakanakaevangeli are doing here to resolve this minibuffer quitting issue! I am not really qualified to judge since as I said I didn't follow the discussion closely. My questioning was simply based on the superficial observation of a complex discussion and the incoming seemingly complicated solutions. 

Generally, from my perspective it is better to not trying to work to hard against the Emacs standard behavior. And the minibuffer does reset the window configuration by design (this is baked into the minibuffer code, see read_minibuf in minibuffer.c). I think I will just adapt to what is given here :smile:

My opinion regarding your points:

1. :heavy_check_mark: The single level behavior should be kept
2. :heavy_check_mark: embark-act with configurable quitting is good and should be kept
3. :heavy_multiplication_x: embark-quit should probably go away

---

I would like to ask one thing - there is the issue #117 where you proposed to restart the minibuffer if an action has been executed. Could it work if we always execute actions outside the minibuffer and always restart? Fixing the window configuration and the refreshing in one go.

**Comment by oantolin** [comment]
> 1. ✔️ The single level behavior should be kept
> 2. ✔️ embark-act with configurable quitting is good and should be kept
> 3. ✖️ embark-quit should probably go away

Just in case my earlier wordiness didn't make it clear, the last commit, 976246cc83b6c619fbc78592252269c7a1103d49, does exactly that: single level quitting, single configurable embark-act command, removes embark-quit.

**Comment by oantolin** [comment]
> Could it work if we always execute actions outside the minibuffer and always restart? Fixing the window configuration and the refreshing in one go.

Maybe. Interesting idea. Although I'm already starting to accept in my heart that window configuration changes for non-quitting actions are ephemeral. 😆

**Comment by jakanakaevangeli** [comment]
Closing this as fixed by 976246cc83b6c619fbc78592252269c7a1103d49. I
will add a wiki entry about embark-quit (C-g), embark-exit (RET), and
advising embark--act to use it instead of abort-recursive-edit, or
perhaps even advising abort-recursive-edit and exit-minibuffer itself.

The whole ordeal was a pretty good exercise for the brain.
